### PR TITLE
[review] feat(pubsub): add support for transportSettings of dataSetReader

### PR DESCRIPTION
**Pull request by waheedejaz**
server_pubsub.h: add extensionObject for transportSettings in DataSetReaderConfig
ua_network_pubsub_mqtt.c: correct datatype checking regist function and QoS mapping
ua_pubsub_reader.c: copy transportSettings when copying dataSetReader_config and some cosmetic  changes

datatypes_pubsub.txt: add BrokerDataSetReaderTransportDataType and JsonDataSetReaderMessageDataType

#2865

### CODE REVIEW

File: src/pubsub/ua_pubsub_reader.c [hunk 1]
```diff
@@ -287,7 +285,7 @@ void UA_ReaderGroup_subscribeCallback(UA_Server *server, UA_ReaderGroup *readerG
         UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Message buffer alloc failed!");
         return;
     }
-
+    buffer.length = 0;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:288]: Not sure about that one.
Isn't the length of the available buffer checked in the `receive` function?

Comment (waheedejaz) [reply src/pubsub/ua_pubsub_reader.c:288]: buffer length is initialized to 0 here.  After the receive() function,  the length of the buffer is checked to determine if a message is received or not. The check can successful due to garbage value even if the message is not received.

**Comment by opcua-tsn-team-kalycito** [comment]
> Looks good to me.
> Please ok this one @andreasebner, @opcua-tsn-team-kalycito .

@jpfr Looks good to us.

File: src/pubsub/ua_pubsub_reader.c (hunk 1 above)
Comment (waheedejaz) [reply src/pubsub/ua_pubsub_reader.c:288]: Upon closer look, this is not correct. The buffer.length is the total size of the buffer, not the size of the received buffer. It is used in memset(0) in allocate and delete buffer functions. 

But it is also used to check the length of the received message on the next line. In fact, the decision, whether or not a message is received,  is made by checking this length. Total allocated length is 512 bytes, so this check would always be true whether or not the message is received.  Currently, MQTT channel doesn't have receive function, code just explodes at this point. Both UDPMP and ETHERNET receive functions change the buffer length to 0 or to the received message size. This also causes memory leaks. We can maybe change the structure of  UA_ByteString, add one more variable for current size. In any case, the handling of pubsub receive must be change.

For now, I am removing this line.

Comment (jpfr) [reply src/pubsub/ua_pubsub_reader.c]: > Both UDPMP and ETHERNET receive functions change the buffer length to 0 or to the received message size. This also causes memory leaks.

That should not happen. The line that sets the length to zero should also free the memory.
Setting to a smaller positive value is okay as `free` does not care in the end.

Can you open separate issues if the above is confirmed?

Comment (waheedejaz) [reply src/pubsub/ua_pubsub_reader.c]: It is being freed at the end of subscriberCallBack() in all cases. I was concerned about the memset(0). Didn't realize that type->size and buffer size are different. 
All good!

**Comment by jpfr** [comment]
The PR does two things:

1. Fixes a few issues around data types.
2. Adds the TransportSettings.

The first looks very good.
For the second one, I see the settings copied over. But they are not deleted at the end of the Reader's lifecycle. Further, nobody is using the TransportSettings at the moment. What are they for?

Is it enough to keep the fixes and use the TransportSettings only during the initialization of the Reader?

**Comment by waheedejaz** [comment]
> I see the settings copied over. But they are not deleted at the end of the Reader's lifecycle. 

TransportSettings deletion added. PR updated.

> Further, nobody is using the TransportSettings at the moment. What are they for?

I am working on MQTT_Subscribe example. Didn't want to put up a lot changes for review at once.

**Comment by jpfr** [comment]
Got it. Small PR are good.
Merged.

**Pull request merged**